### DISCUSSION TITLE
How to achieve completely manual endpointing?

### DISCUSSION THREAD

**Discussion by antxv** [deepgram/community#821] [General help] [answered]
I am the lead developer for [Utterly Voice](https://utterlyvoice.com/), which is a voice dictation application designed for people with hand issues. It supports multiple recognizers, including Deepgram. I am attempting to update the client code to use the latest Go SDK, which bundles all websocket calls within the client. I have tried various values of `Endpointing`, `InterimResults`, and `UtteranceEndMs`, but some amount of automatic endpointing always occurs. Many of the utterances are commands with arguments, so I need absolute control over defining the end of utterance. I am attempting to use the `Client.Finalize()` function when our application detects the end of utterance, but Deepgram does not seem to honor that. Is there any way to get complete control over end of utterance? Without this ability, we will need to remove Deepgram as an option for our voice dictation application.

In case it matters, the audio our application sends is in chunks varying from 1-30 seconds each. The application starts streaming the utterance when it gets detected from the microphone. When the application detects that the utterance is complete, it stops streaming data and calls `Client.Finalize()`. We cannot continuously stream all microphone data, because our users typically run the application for several hours every day, and this would create a massive bill. We need to stream each utterance, one at a time, and get one finalized transcript for each utterance.

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by nikolawhallon** [comment, marked as answer]
You might want to start by setting `endpointing=false` and `interim_results=false` (and don't use `utterance_end_ms`) to remove those algorithms firstly to not have to think about them for now.

After that, Deepgram will still need to make chunks itself - if you stream 30 seconds of audio, Deepgram will need to chunk that audio up to inference it to stream back the transcripts. Internally, it makes chunks every 3-5 seconds.

But what you can do to accomplish what you are looking for is to concatenate all transcripts you receive, then when you determine the end of the utterance, do `Client.Finalize()`, continue concatenating transcripts until you receive a message with the field `from_finalize` set to `true` - at that point, you should have the full transcript of the full utterance up until the point in the audio sent when you called `Client.Finalize()`.

I do this in my own projects, and it works well, here's some pseudo-code from my project:

```
fn on_audio(audio) {
  ...
  dg_ws.send(audio)
  ...
}

fn on_speech_ended() {
...
  dg_ws.send(r#"{ "type": "Finalize" }"#)
...
}

full_transcript = ""

fn on_dg_message(dg_message) {
  full_transcript += dg_message.transcript
  if dg_message.from_finalize {
    print("the full utterance was: " + full_transcript)
    // reset the transcript for the next utterance
    full_transcript = ""
  }  
}
```

**Reply by antxv** [reply]
Thanks! I was missing the `FromFinalize` part of that process. I have one further complication with this. I am setting the number of alternatives to 3 in the request, so I can apply biasing of transcripts within my application (for example, to bias in favor of command names within a transcript). If I send one utterance that gets broken up into two chunks, I get 3 alternatives, then another 3 alternatives. Can I safely concatenate the ith alternative from each of those chunks? I am just not certain if there is any guarantee of alternate ordering consistency from one chunk to the next.

**Reply by nikolawhallon** [reply]
> Can I safely concatenate the ith alternative from each of those chunks?

my understanding is "yes" but also that there may be better ways to achieve what you are looking for in alternatives, such as keyword boosting

**Reply by antxv** [reply]
Thanks, that is how it appears to behave on testing. I am just hoping I am not getting lucky during testing :-) The bias I apply in my application is more complicated than keyword boosting. It is based on regular expressions and I sometimes have more than the limit of 100 keywords allowed by the API.
